### ISSUE TITLE
How to convert frames' byte[] (provided by setFrameCallback) to org.webrtc JavaI420Buffer?

### ISSUE THREAD

**Issue by acrolink** [jiangdongguo/AndroidUSBCamera#218]

**Comment by konovalov-k** [comment]
Hi again @acrolink, just use as 
`private val myFrameCallBack: IFrameCallback = IFrameCallback { frame ->
        val videoFrame = VideoFrame(
            NV12Buffer(
                width,
                height,
                width,
                height,
                frame,
                { JniCommon.nativeFreeByteBuffer(frame) }), 0, System.nanoTime() - start
        )
        capturerObserver.onFrameCaptured(videoFrame)
    }
    camera.setFrameCallback(myFrameCallBack, UVCCamera.PIXEL_FORMAT_NV21)  
`

**Comment by konovalov-k** [comment]
https://gist.github.com/konovalov-k/a8e834465fc26e7fd67c96d6b3af1513

**Comment by acrolink** [comment]
@acrolink 

Thank you. I have managed to make it work on the rate of 95%. The picture is fine but sometimes the image gets green or distorted.

What is working for me:
`camera.setFrameCallback(mIFrameCallback, UVCCamera.PIXEL_FORMAT_YUV420SP);`
and:
```
public void addFrame(ByteBuffer frame) {
 
            byte[] imageArray = new byte[frame.remaining()];
            frame.get(imageArray);
            frame.rewind();

            NV21Buffer nv21Buffer = new NV21Buffer(imageArray,640,480, null);
            VideoFrame videoFrame = new VideoFrame( nv21Buffer, 0, System.nanoTime() );
            capturerObs.onFrameCaptured(videoFrame);
}
```
Correct me if I am doing something wrong.

**Comment by konovalov-k** [comment]
About distorted / green image... I don't know why. I got the same problem.
Is PIXEL_FORMAT_YUV420SP work well with NV21 buffer? I had some problems in past with that stuff. 
Also why you don't provide lambda for clear buffer on native side?

**Comment by acrolink** [comment]
Could you please give an example of using a lambda in this context.

**Comment by konovalov-k** [comment]
As I did here NV12Buffer( width, height, width, height, frame, { JniCommon.nativeFreeByteBuffer(frame) }), 0, System.nanoTime() - start )

**Comment by acrolink** [comment]
I have changed my code to this:

```
  public void addFrame(ByteBuffer frame) {
            long start = System.nanoTime();
            byte[] imageArray = new byte[frame.remaining()];
            frame.get(imageArray);
            NV21Buffer nv21Buffer = new NV21Buffer( imageArray, 640, 480, () -> JniCommon.nativeFreeByteBuffer(frame));
            VideoFrame videoFrame = new VideoFrame( nv21Buffer, 0, System.nanoTime() );
            capturerObs.onFrameCaptured(videoFrame);
}
```
Now the issue of green and corrupted frames happens less frequently but I guess to eliminate it completely I may have to validate the `videoFrame` as valid (good) before sending it over to `capturerObs.onFrameCaptured()`. Any ideas regarding this?

**Comment by konovalov-k** [comment]
Maybe get bytes from U and V channel, mix it and if in random places it green don't send?)

**Comment by konovalov-k** [comment]
@acrolink And i think it is not the issue of this lib :)

**Comment by acrolink** [comment]
> Maybe get bytes from U and V channel, mix it and if in random places it green don't send?)

Could you kindly refer me to some tutorial online on how to do that?

**Comment by konovalov-k** [comment]
It is just idea based on colour theory. I don't have tutorial or solution too. :(

**Comment by acrolink** [comment]
I suspect that the corrupted and green frames issue is related to WebRTC's `VideoFrame` itself since the same frames provided by UVC's `iframeCallback` when passed to OpenTok's or Agora's API produce perfect video play.

**Comment by konovalov-k** [comment]
It may be wrong, because i use this lib in other cases and still getting green images from time to time.

**Comment by pumatus** [comment]
@konovalov-k Hi~ Is this problem resolved?

**Comment by konovalov-k** [comment]
Hi @pumatus. Main problem with convertation or with green frames from time to time?

**Comment by pumatus** [comment]
@konovalov-k  Hello~ I want to send the uvc preview data (byte[]) to webrtc, there is a problem here, how do you deal with it?

**Comment by konovalov-k** [comment]
@pumatus Yeah, we solve it. In my projects I use standart Google WebRTC lib. And it use VideoCapturer class to provide frames from abstract camera and WebRTC. I wrote [this](https://gist.github.com/konovalov-k/a8e834465fc26e7fd67c96d6b3af1513) to provide connection between USBCamera and WebRTC. Hope it helps.

**Comment by pumatus** [comment]
@konovalov-k Thank you very much~ 大佬～

**Comment by heaveninfosoft** [comment]
@acrolink  can you send me code??

**Comment by acrolink** [comment]
@heaveninfosoft 

Hope this helps:

```
    void addFrame(byte[] frame) {
        try {
            final long captureTimeNs =
                    TimeUnit.MILLISECONDS.toNanos(SystemClock.elapsedRealtime());
            VideoDimensions dimensions = new VideoDimensions(camera_WIDTH, camera_HEIGHT);
            VideoFrame videoFrame = new VideoFrame(frame,
                    dimensions, VideoFrame.RotationAngle.ROTATION_0, captureTimeNs);
            videoCapturerListener.onFrameCaptured(videoFrame);
        } catch (Exception ignored) {
        }
    }
```

**Comment by heaveninfosoft** [comment]
@acrolink please send me full code 
VideoDimensions ,VideoFrame  ,videoCapturerListener  what is all this???? where is main lib?

**Comment by heaveninfosoft** [comment]
@pumatus  hi send me full code

**Comment by pumatus** [comment]
I didn't figure it out. I gave up... - -! @heaveninfosoft

**Comment by heaveninfosoft** [comment]
@acrolink please send me full code
VideoDimensions ,VideoFrame ,videoCapturerListener what is all this???? where is main lib?
